### DISCUSSION TITLE
Smooth video or flythrough of the data

### DISCUSSION THREAD

**Discussion by Pietersielie** [idia-astro/iDaVIE#406] [Upcoming features]
Currently, recording the user's view of the data is done through recording the screen showing the view of one of the user's eyes. This video is invariably jittery (because of the user's head movement) and not of good quality. Recording a video from a Unity camera moving through the data results in smoother and better quality video. A prototype was created that utilised a hard-coded route for the camera to follow. It is desirable to create a way for the user to control the movement of this camera, without sacrificing the quality. One way to do this would be by having a script with commands that the camera will follow, and a button on the desktop interface to have it execute. Alternatively, the user could control the camera using buttons to direct movement, orientation, and zoom, though this will require concentration from the user. A potential solution is to use the keyboard controls to mark waypoints, and then have the automated camera navigate between the waypoints. See the [relevant](https://github.com/idia-astro/iDaVIE/issues/132) [issues](https://github.com/idia-astro/iDaVIE/issues/240) development tracking.

**Comment by maystey** [comment]
I am developing this feature in the [mayhew/VideoMaker](https://github.com/idia-astro/iDaVIE/tree/mayhew/VideoMaker) branch. While much of this feature is still a work in progress, I will describe the current design proposal below, created in collaboration with @Pietersielie and @Michaelazzz. If you have any questions or suggestions, please feel free to add to the conversation.

### Video Maker Design Proposal

In this design proposal I will focus on using iDaVIE's desktop mode. VR integration will be considered later. 

In the desktop editor a tab called "Video" will be added. In this tab the user can load a VideoScript (a json file with settings and commands to move the camera), preview the video (with a viewport that shows the camera texture) and export the video. The video will be exported to the "Output" folder directory (same root folder as screen captures).

In the short term we will define an API in the documentation for the VideoScript file, as well as a robust parser that will provide error messages to the user if there is something wrong in the file. In the medium-term we plan to include a dedicated editor for the VideoScript, possibly as a stand-alone web application (due to the limitations of the Unity UI framework).

As stated earlier the VideoScript has a sequence of actions which I will refer to as VideoActions/Transitions (data structures). These define the motion and viewing angles / targets of the camera over time. Each VideoAction/Transition is given a duration and will execute subsequently with no overlaps. I will discuss these in detail later.

The parsed VideoScript will be sent to the CameraActionConstructor, which is responsible for generating 3 queues of CameraActions (classes), responsible for changing the position, look direction and up direction of the camera during runtime. These CameraActions are more granular than the VideoActions/Transitions and are focused on performing a single task; a single VideoAction/Transition may generate mutliple CameraActions in each queue.

When previewing or exporting the video file, the CameraAction qeueues are sent to the CameraController, which is responsible for setting the camera transform using the results from the CameraActions. As the CameraController updates it will run through the CameraAction queues, discarding CameraActions that are completed (when the duration is reached) and using the next CameraActions in the queues. The CameraController is also responsible for capturing frames to be rendered into the exported video.

The video will be created using FFmpeg, either by invoking the executable directly, or by using the [FFmpegOut](https://github.com/keijiro/FFmpegOut) or [Spirit30/FFmpegUnityBind2](https://github.com/Spirit30/FFmpegUnityBind2) Unity packages. I am still looking into which of these options are best for our use-case. Note that the Unity Recorder addon only works in the Unity Editor and not in stand-alone builds (which is required for this feature).

That concludes the pipeline for the VideoMaker feature. In the following sections I will discuss some of the classes / data structures in more detail.

#### Supporting Classes

The CameraActions will rely on some supporting classes that I will discuss in this section. These classes are EasingFunctions and Paths.

EasingFunctions are generally used in animations to control the timing of transitions / movements. The EasingFunction class provides a method that maps a normalised input parameter to a normalised output parameter.

Some examples of EasingFunctions are:
- Easing in - 0 gradient at 0, polynomial order can be defined (starting at 2)
- Easing out - 0 gradient at 1, order can be defined
- Easing in/out - easing in from 0 to 0.5 and easing out from 0.5 to 1

The Path class defines a parameterised curve in space. The curve parameter will be normalised. The Path class has an optional property for an EasingFunction to be set, which will be applied to the curve parameter. The base Path class defines 3 methods for derived classes to override; each take the curve parameter as an argument and return a vector:

- Position - position along the path
- Tangent - **normalised** tangent along the path
- Curvature - **normalised** curvature along the path

Some examples of derived Path classes include:
- A straight-line path
- A circle path, for which multiple rotations and partial arcs can be defined
- A spiral path (perhaps a more general circle path)
- A cubic spline path

#### CameraActions

The CameraActions are used to define the transform of the camera over time. The only property required by the base CameraAction class is its duration, so that the CameraController knows when an action is complete. Note that there are to be no gaps in time in the CameraAction queues, thus if the camera's position or orientation needs to stay constant an action must still be used for that.

The three CameraAction queues are specifically used to define the position, look-at direction and up direction of the camera over time (all vectors). To support this, there are two major types of CameraActions: CameraPositionAction (which has a method to get the position at a given time) and CameraDirectionAction (which has a method to get the direction at a given time and position). The reason the CameraDirectionAction requires the current position to give the direction is for cases where a look-at target is defined.

Some examples of CameraPositionActions are:
- CameraPositionActionHold - the camera's position is constant
- CameraPositionActionPath - the camera position follows a given Path

Some examples of CameraDirectionActions are:
- CameraDirectionActionHold - the direction is constant
- CameraDirectionActionTween - the direction transitions from one direction to another over the duration, with an optional EasingFunction to control the timing
- CameraDirectionActionLookAt - direction points to a target
- CameraDirectionActionPath - the direction points in or away from either the tangent or curvature direction of a given Path
- CameraDirectionLookAtPath - the direction points to the position along a given Path

#### VideoScript

The VideoScript will consist of settings for the video, including:
- Resolution
- Frame rate
- iDaVIE logo position and size
- Render settings? (v_min/max, colormap, etc)

as well as a sequence of VideoActions/Transitions. VideoActions can be used to define positions and viewing targets for the camera, and VideoTransitions can be used to smoothly move the camera from the previous VideoAction to the next. Although the standard VideoAction should be sufficient to describe most camera motions desired, more specific VideoActions will be defined for convenience, for example VideoActionOrbit which defines an action to orbit a given target.

I will return to elaborate on this section later.

**Comment by maystey** [comment]
### Creating Videos with FFmpeg

As mentioned in the design proposal, there are different options for creating the video. FFmpeg seems to be the best bet as the Unity Recorded is an option. We can implement this ourselves or use existing Unity Packages:
- [FFmpegOut](https://github.com/keijiro/FFmpegOut)
- [FFmpegUnityBind2](https://github.com/Spirit30/FFmpegUnityBind2)

I am hesitant to use these packages for various reasons. One is that they don't seem to be very widely adopted or actively developed. Another is that they offer general solutions that are beyond the scope of what we require and seem to be focused on real-time recording (whereas we can choose the framerate and render each frame we need).

Shipping FFmpeg binaries with iDaVIE poses licensing concerns, as some of the FFmpeg binaries are distributed under the GPL, a stricter license than iDaVIE. This can be circumvented by requiring that the user installs FFmpeg themselves and adds a path to the executable in a config file if they want access to video capture.

**Reply by Pietersielie** [reply]
My preference would be to require FFMPEG to be installed should the user wish to create videos. It is not difficult to install, and we can add a popup message if FFMPEG is not installed (i.e., the config file does not have a FFMPEG location, or the location does not exist). This message then prompts the user to choose a FFMPEG executable, or provide a link for download.

**Reply by vdhulst** [reply]
I fully agree. It also ensures that people using the prebuilt package can use this option. Most will have FFMPEG already and if not: it is easy to install. 

It should be well documented anyway.

**Comment by maystey** [comment]
### VideoScript Proposal

Here I present the proposed structure of the Video Script json. If you have any questions or comments about it, please feel free to add to the conversation.

A VideoScript uses the standard Json format, with an object (dictionary) as the root. Note that properties marked as **optional** or with a **default** value do not need to be defined in the script. The properties of the VideoScript are:

- width : int

    Width of the image in pixels.

- height : int

    Height of the image in pixels.

- frameRate : int

    Frames per second of the video.

- useLogo : bool (default true)

    If true use the iDaVIE logo (perhaps this shouldn't be an option).

- logoPosition : string (defualt "bottomRight")

    The (named) position of the logo, with vertical part "top"/"center"/"bottom" and horizontal part "Left"/"Center"/"Right"

- logoX : float (optional)

    The normalised x coordinate for the logo position on screen (origin in the bottom left)

- logoY : float (optional)

    The normalised y coordinate for the logo position on screen (origin in the bottom left)

- logoSize : float (defualt 0.2)

    The size of the logo as a fraction of the screen width.

- vars : object

    Properties of this object are variables that can be used in defining subsequent variables, and in the actions.

- actions : array of objects

    An array of action or transition objects.

#### VideoScript Objects

All of the VideoScript objects have a `type` property with a string value. Note that the Variable type can be used for any property as it is a proxy for other values / objects.

> **Variable** (type: "var")
> A reference to a value or object defined in `vars`.
> 
> Properties:
> - name: string
>  
>     Property name used in `vars`.


> **Direction** (type: "direction")
> A direction vector defined in 3D space. These are automatically normalized. Zero vectors will be changed to a "forward" direction, or a default position defined by the context in which the direction is used.
> 
> In most contexts when a property requires a Direction object, it will also allow for string values. These are for named directions, which can be used instead of a direction object. The named directions are:
> - "forward" - facing the front face of the cube (0, 0, 1)
> - "back" - (0, 0, -1)
> - "left" - (-1, 0, 0)
> - "right" - (1, 0, 0)
> - "up" - (0, 1, 0)
> - "down" - (0. -1, 0)
>
> Properties:
> - x : float (default 0)
> - y: float (default 0)
> - z: float (default 0)


> **Position** (type: "position")
> A position in 3D space. This can be defined in engine coordinates x/y/z, WCS coordinates ra/dec/freq/velo/... or relative to a source id from the mask (center position using engine coordinates or wcs coordinates as an offset).
>
>  In most contexts when a property requires a Position object, it will also allow for string values. These are for named positions, which can be used instead of a Position object. The named positions are:
> - "center" - The center of the data cube
>
> Properties:
> - x : float (defualt 0)
>
>    The x component of the position in engine coordinates.
>
> - y : float (defualt 0)
>
>    The y component of the position in engine coordinates.
>
> - z : float (defualt 0)
>
>    The z component of the position in engine coordinates.
>
> - ra : float (optional)
>
>    The Right Ascension of the position in WCS.
>
> - dec : float (optional)
>
>    The Declination of the position in WCS.
>
> - freq / velo / vrad / vopt / wave / zopt : float (optional)
>
>    The component of the position along the spectral axis in WCS.
>
> - sourceId : int (optional)
>
>   Source ID from the mask to make the position relative to.


> **Easing** (type: "easing")
> An easing function with generic options. Named easing functions may include options that are not covered by generic properties.
>
> Using generic options, the easing function is an ease in/out function with a linear middle. The time span for each of these parts can be defined.
>
> Properties:
> - name : string (optional)
>
>    Which standard easing function to use. Options are:
>    - "in" - easing in
>    - "out" - easing out
>    - "inOut" - ease in and out
>
> - order : int (default 2)
>
>    The order of the easing functions (slope). Default of 2 is quadratic.
>
> - timeIn : float (default 0.5)
>
>    The relative time for the (starting) ease-in phase of the easing function.
>
> - timeMid : float (defualt 0)
>
>    The relative time for the (middle) linear phase of the easing function.
>
> - timeEnd : float (optional)
>
>    The relative time for the (ending) ease-out phase of the easing function. Only two times should be defined.


> **Path** (type: "path")
> A path in 3D space. All paths can be defined using a start and end position, but these properties are sometimes optional. 
>
> As the different types of paths have their own properties, these are presented in their own sub-blocks below the general properties. General properties are relevant for all path types.
>
> Properties:
> - name : string (default "line")
>
>    Which path to use. Options are:
>    - "line" - straight line path
>    - "circle" - circle path
>    - "spiral" - spiral path
>    - "ellipse" - elliptical
>    - "spline" - cubic spline path
>
> - easing : Easing (optional)
>
>    The easing function to control the timing of the path parameter.
>
> - startPosition : Position (optional)
>
>    The start position of the path. Can be defined for all path types.
>
> - endPosition : Position (optional)
>
>    The end position of the path. Can be defined for all path types.
>
> > **Circle Path** (name: "circle")
> > The Circle path has many optional properties, but to define the path properly the following combinations are required:
> > 1. `centerPosition`, `startPosition` and `endPosition` (`rotations` is optional, but treated as an integer)
> >
> >    In this case the distance between the `centerPosition` and `startPosition` is used to determine the radius of the circle. If the distance between the `endPosition` and `centerPosition` is not the same as the radius, the final position of the path will not match the `endPosition` exactly. If a change in radius is required, use a Spiral Path instead.
> >
> >    The `rotations` property can be used to define additional full rotations before ending at the `endPosition`.
> > 
> > 2. `centerPosition`, `startPosition` / `endPosition`, `axis` and `rotations`
> > 3. `centerPosition`, `axis` (named direction only), `startAngle` / `startAngleDeg` / `startAngleRad`, `endAngle` / `endAngleDeg` / `endAngleRad`, `radius` (`rotations` is optional, but treated as an integer)
> > 
> >     Whenever an initial or final angle is defined a zero-angle reference point is needed. This is difficult to define for circles with any possible axis direction, thus only axes with named directions are supported. See the definition of `startAngle` in the properties below for the definitions of the zero-angle positions.
> >
> >     The `rotations` property can be used to define additional full rotations before ending at the defined end angle.
> >
> > 4. `centerPosition`, `axis` (named direction only). `startAngle` / `startAngleDeg` / `startAngleRad` / `endAngle` / `endAngleDeg` / `endAngleRad`, `rotations` / `arcDeg` / `arcRad`, `radius`
> >
> > Properties:
> > - centerPosition : Position
> >
> >    The position of the origin of the circle / spiral / ellipse.
> >
> > - axis : Directional (default "up")
> >
> >    The direction of the circle's axis. The path direction around the axis follows the right-hand rule.
> >
> > - radius : float (optional)
> >
> >    The radius of the circular path.
> >
> > - startAngle : float (optional)
> >
> >    The starting angle (fractional) of the path. This can only be used if `axis` is defined as one of the named directions. The 0 angle position is defined as on of the following, depending on the axis direction:
> >    - "up/down" - 0 angle on the -z-axis
> >    - "forward/back" - 0 angle on the +x-axis
> >    - "left/right" - 0 angle on the +y-axis
> >
> > - startAngleDeg : float (optional)
> >
> >    The starting angle in degrees. Equivalent to `startAngle`.
> >
> > - startAngleRad : float (optional)
> >
> >    The starting angle in radians. Equivalent to `startAngle`.
> >
> > - endAngle : float (optional)
> >
> >    The ending angle (fractional). The same requirements and angle position as `startAngle` apply.
> >
> > - endAngleDeg : float (optional)
> >
> >    The ending angle in degrees. Equivalent to `endAngle`.
> >
> > - endAngleRad : float (optional)
> >
> >    The ending angle in radians. Equivalent to `endAngle`.
> >
> > - rotations : float (optional)
> >
> >    Number of rotations. This can be fractional for partial rotations.
> >
> > - arcDeg : float (optional)
> >
> >    Arc covered by the circular path in degrees, may exceed 360 for multiple rotations.
> >
> > - arcRad : float (optional)
> >
> >    Arc covered by the circular path in radians, may exceed 2pi for multiple rotations.
> 
> > **Spiral Path** (name: "spiral")
> > The Spiral Path is like a Circle Path, but with an increasing / decreasing radius. The combinations of properties required for the Spiral Path is the same as the Circle Path, except:
> > - For combination 1. the distance between `startPosition` and `centerPosition` is used for the starting radius and the distance between the `endPosition` and `centerPosition` is used for final radius.
> > - For combinations 3. and 4. `radius` should be replaced with 2 properties: `startRadius` and `endRadius`. If only `radius` is given the path will be circular.
> >
> > Properties:
> > - startRadius : float (optional)
> >
>>    The starting radius of the spiral. Synonymous with `radius`. 
> >
> > - endRadius: float (optional)
> >
> >    The ending radius of the spiral.
>
> > **Ellipse Path** (name: "ellipse")
> > **WIP**
>
> > **Spline Path** (name: "spline")
> >
> > Position:
> > - knots : array of Positions ... **WIP**


> **Action** (type: "action")
> A standard video action that defines the position and orientation of the camera over time.
>
> Properties:
> - duration : float
>
>    Duration of the action in seconds
>
> - position : Position / Path (optional)
>
>    The position of the camera. A static position if a Position object is given, or the position along the given Path object.
>    If not define, the last position from the previous action is used.
>
> - lookAt : string / Direction / Position / Path (optional)
>
>    The direction in which the camera is looking. Named directions or a Direction object can be used to define a static direction for the camera over the duration of the action. Other options are described below.
>
>    Viable string values are:
>    - Named directions "forward" / "back" / "left" / "right" / "up" / "down"
>    - Named positions "center"
>    - "alongPath" - look along the tangent of the position Path
>    - "alongPathReverse" - look in the opposite direction to the tangent of the position Path
>    - "upPath" - look along the curvature direction of the position Path
>    - "upPathReverse" - look in the opposite direction to the curvature of the position Path
>
>    If a Position is given, then the camera will look at this position over the duration of the action. If a Path is given, then the camera will look at the position along the path over the duration.
>
>    If not defined and the position is static, then the last lookAt direction of the previous action is used. If the position is a Path, then lookAt defaults to "alongPath"
>
> - lookUp : string / Direction / Position / Path (optional)
>
>    The "up" direction used when determining the camera orientation. This is not strictly the up direction of the camera, as this would require the lookAt and lookUp directions to be orthogonal.
>
>    Same options as lookUp.
>
>    If not defined and the position is static, then last lookUp direction of the previous action is used. If the position a Path, then lookUp defaults to "upPath".


> **Transition** (type: "transition")
> Can be placed between Actions to ensure that the camera seamlessly transitions between the preceding to proceeding actions.
> Transition has 3 optional properties which define consecutive phases of the transition: lookStart, travel and lookEnd.
>
> Properties:
> - lookStart : object (optional)
>
>    At the start of the transition, rotate the camera from the ending direction of the preceding action to the starting direction of the travel phase (if included) or the starting direction of the next action.
>
>     Properties:
>    - duration : float
>
>        The duration of the lookStart phase in seconds.
>
>    - easing : Easing (optional)
>
>        Optional easing function to re-time the transition from the starting orientation to ending orientation.
>
>     - overlapWithTravel : bool (default false)
>
>        If true, then this phase overlaps with the travel phase (final camera direction matches the travel direction at the lookStart duration).
>        Note that the combined duration of overlapping lookStart and lookEnd phases cannot exceed the duration of the travel phase.
>
> - travel : object (optional)
>
>    Move the camera from the ending position of the previous action to the starting position of the next action.
>
>    Properties:
>    - duration : float
>
>        The duration of the travel phase.
>
>    - easing : Easing (optional)
>
>        Optional easing function to be applied to the default line path (if a path is not defined for `position`).
>
>    - position : Path (optional)
>
>        Path for the camera to move over during the travel phase. If none specified, then a default line path is used.
>        Note that the startPosition and endPosition properties of the path do not need to be defined as the ending position of the previous action and the starting position of the next action will be used as defaults respectively.
>
>    - lookAt : string / Direction / Position / Path (optional)
>
>        The direction for the camera to look at during the travel phase. The options and defaults are the same as the `lookAt` property of Action.
>
>    - lookUp: string / Direction / Position / Path (optional)
>
>        The up direction for the camera during the travel phase. The options and defaults are the same as the `lookUp` property of Action.
>
> - lookEnd : object (optional)
>
>   At the end of the transition, rotate the camera from the ending direction of the travel phase (if included) or the ending direction of the previous action to the starting direction of the next action.
>
>    Properties: See properties of `lookStart` phase.


> **ActionOrbit** (type: "actionOrbit")
> An action specifically designed for orbits / turntables, easier to use than the standard action. 
> **This is a WIP, as the circle path needs some revision.**

#### Examples

**Turntable**
```
{
    "width": 1920,
    "height": 1080,
    "frameRate": 30,
    "actions": [
        {
            "type": "action",
            "duration": 15,
            "position": {
                "type": "path",
                "path": "circle",
                "centerPosition": "center",
                "radius": 2,
                "startAngle": 0,
                "rotations": 1
            },
            "lookAt": "center",
            "lookUp": "up"
        }
    ]
}
```

**Reply by maystey** [reply]
Based on internal discussions we are moving away from Json to use a simpler built-for-purpose scripting language.

**Comment by Pietersielie** [comment]
As mentioned previously, we will move the front end scripting to a very simplified, for-purpose scripting language to promote useability for less technically inclined users, rather than JSON. The backend will remain as described, with improvements and additions as they are developed. We provide the writeup below as a very early prototype, with significant changes likely to happen based on feedback from users.

For the first prototype, we envisage that the user would be able to select a series of positions within the VR space, clicking a button that then records the position of the user's head, relative to the data cube (both position and scale). This will be in a separate mode akin to the painting mode, accessed from the quick menu. The user can press a button on the controller to record a position. Each position consists of a **location** within the engine space, i.e., `(X, Y, Z)`, and an **orientation**, called the rotation in Unity, `(x, y, z)`. This preserves the viewpoint. Alternatively, the orientation might consist of two Unity engine vectors relative to the user's head, called `up` and `forward`, i.e., the local y-axis and the local z-axis respectively. Whether the rotation or the vectors are used for orientation will be fixed in later iterations as determined by user tests and the requirements from the backend.

This list of positions can then be exported to a file outside of the VR space. This file will have some formatting already to make it suitable as a script to edit and load in the desktop, with numeric aliases (of the form `pN`) provided for each position. The list of positions are treated as a FIFO queue: After the first, initial position is determined, any future instruction will only contain the destination, i.e., the start position is always the current position.

For the first prototype, we envisage having three possible commands.
- First, move to a position over a user specified time period, with a few different movement options such as a straight line or an arc.
  - This takes the form `Move in METHOD to POSITION over N seconds`.
  - Future iterations will allow customisation of the easing function used, but for the first prototype, a sigmoid-like easing function called `InOut` (see [above](https://github.com/idia-astro/iDaVIE/discussions/406#discussioncomment-13876792)) will be used by default.
- Secondly, rotate around a given position on a turntable, for a specified number of turns, with optional speed control. The camera will look at the centre point throughout the orbit.
  - This takes the form `Rotate around POSITION N times`.
  - Optionally, `turn Q seconds` and/or `orbit R seconds` can be added to this command.
    - These determine the speed at which the camera rotates to face the centre point of the turn table, and how long an orbit takes to complete.
    - These options have a default value that can be set in the configuration file, likely to be set at 10 seconds for a complete rotation by default.
  - The radius is determined from the starting position, i.e., the current position in the queue.
  - For now, rotation is around the up axis (of the world space). This will be customisable in future iterations.
  - `N` is a floating point number, allowing for half or quarter rotations.
  - E.g., in the example below, the rotate command starts at `p3`, looks at `p4`, and rotates 3 times around the world-space `up` vector passing through `p4`.
- Lastly, a wait command to pause before moving to the next action.
  - This takes the form `Wait N seconds`.

An example script could be as follows, assuming the positions are to zoom in, move partway around it, rotate, move back and zoom out:
```
# List of positions:
----------
p1 is {[X1,Y1,Z1],[x1,y1,z1]}
p2 is {[X2,Y2,Z2],[x2,y2,z2]}
p3 is {[X3,Y3,Z3],[x3,y3,z3]}
p4 is {[X4,Y4,Z4],[x4,y4,z4]}

# Script:
----------
Start at p1
Wait 1 seconds
Move in STRAIGHT to p2 over 4 seconds
Move in ARC to p3 over 2 seconds
Wait 3 seconds
Rotate around p4 3 times
Wait 2 seconds
Move in ARC to p2 over 4 seconds
Move in STRAIGHT to p1 over 4 seconds
Wait 1 seconds
```

Some method of commenting will be allowed in the script, likely a simple line comment using an octothorpe (`#`) to indicate that anything following that symbol is a comment and to be disregarded.

Future features could more detailed speed control on movements, with easings and more, and hopefully, a way to do this from within the VR environment. A possible improvement on the VR selection of the points is to include an option to use voice recognition to generate user specified comments for each point.

**Reply by veggiesaurus** [reply]
does "rotate around p4" imply that we're at p3, looking at p4 and then rotating around the up axis?

**Reply by Pietersielie** [reply]
Yes, thanks, that's a clarification that got missed in this iteration of the writeup. I've made a change to reflect that.

**Reply by maystey** [reply]
Little nitpick. Would "LINE" be an appropriate replacement for "STRAIGHT" paths? It's just easier to type.

**Reply by Pietersielie** [reply]
Yes, the methods were just placeholders until the backend side is finalised.

**Comment by maystey** [comment]
### JSON Schema
**I still need to test this schema with some test examples. I have also not implemented this schema into the VideoMaker branch yet .  The schema is a WIP and I may edit this comment and upload new files.**

Here is a WIP schema for the Video Script JSON file: 
[video-script-schema.json](https://github.com/user-attachments/files/21976013/video-script-schema.json)


There are parts / options missing at the moment, but the schema should still work. I have also made changes from my previous comment on this. 

Notable changes are:
- Introduction of options "continue", "previous", "next" to refer to next / previous actions.
- Introduction of the Rotation type to better support the front-end scripting langauge.

Parts missing:
- Position from WCS, cube coordinates and sample (long term features)
- Additional options for the circle and spiral paths
- Ellipse and spline paths (long term features)
- "transition" action type. This is not necessary anymore given the changes I have made, though there are aspects of this that may be more convenient, so I may re-introduce it.

### Script Parse Example

Here is an example of a JSON script generated from the scripting-language example in the comment above:
[translated-script-example.json](https://github.com/user-attachments/files/21900016/translated-script-example.json)

Note that I have added comments using `//`, though these aren't valid in a JSON file. There are also certain invalid fields in the JSON file that either represent unspecified values in the example, default values that need to be set or values that need to be determined algorithmically.

I have mostly not relied on default values when writing this, though assuming the defaults can cut things down quite a bit. I thought I would be more stable not to rely on defaults too much this early on, though let me know if you'd like a version that makes use of more defaults.

Relevant to this is the default easing functions (if no "easing" property is provided). I would like to make the default easing function linear (effectively no easing at all), but I believe @Pietersielie would prefer it to be in-out. If that is the general preference, then I will concede.

Note that for the "move in ARC" command I have opted to use a Spiral path. We still need to determine a nice way of determining the center of the arc, and I figured having a variable radius may be necessary. If this isn't the case, then substituting "spiral" with "circle" for the "kind" property is all that needs to change.

**Reply by maystey** [reply]
For the `move in ARC` command, a cubic polynomial (one segment spline) path can be used instead. The full spline will take a bit of work, but a single segment spline should be easy to implement. 

The forward directions of the start and end coordinates can be used to define the second derivatives of the spline boundaries (negate the forward directions if they are looking away from each other). The directions may need to be scaled, perhaps by the (linear) distance between the points divided by duration squared.

**Comment by maystey** [comment]
### Features in RC 1

Release candidate 1 will feature the IDavie Video Script (IDVS) which is used to define the camera actions for the video. We also aim to have VR interactions for selecting and exporting coordinates / rotations for the IDVS to use.

As there are currently issues with the JSON version of the Video Script, this will be excluded from RC1 and will be added later after refactoring. The main issue is that the free version of the Newtonsoft.Json.Schema package limits the number of JSON schema validations per hour, and I was making use of many validations per JSON script when instancing the camera actions. The best alternative package is JsonSchema.Net, which I am having trouble installing for the current version of .net that Unity 2021 uses - so this may have to wait until we have upgraded to Unity 6. I am currently working on refactors to replace the need for an external package for validation, as well as improving the overall design of the code (making it more modular), but this is will not be ready on time.

**Comment by maystey** [comment]
### Features to Add After RC1

These are features that are not a priority for RC 1. Some are large, some are small, quality-of-life features.

- Logo padding (perhaps 10% of the original size)
- JSON video-script format
- Disable VR view camera when previewing video
- Camera debug model in VR when previewing
